pkg/virt-operator/resource/apply/reconcile.go

```diff
@@ -125,15 +132,57 @@ const (
 )
 
 // Merge all Tolerations, Affinity and NodeSelectos from NodePlacement into pod spec
-func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev1.PodSpec) {
+func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev1.PodSpec, nodePlacementOption DefaultInfraComponentsNodePlacement) {
 	if podSpec == nil {
 		podSpec = &corev1.PodSpec{}
 	}
+
 	if componentConfig == nil || componentConfig.NodePlacement == nil {
-		componentConfig = &v1.ComponentConfig{
-			NodePlacement: &v1.NodePlacement{},
+		switch nodePlacementOption {
+		case AnyNode:
+			componentConfig = &v1.ComponentConfig{NodePlacement: &v1.NodePlacement{}}
+
+		case RequireControlPlanePreferNonWorker:
+			componentConfig = &v1.ComponentConfig{
+				NodePlacement: &v1.NodePlacement{
+					Affinity: &corev1.Affinity{
+						NodeAffinity: &corev1.NodeAffinity{
+							RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+								NodeSelectorTerms: []corev1.NodeSelectorTerm{
+									{
+										MatchExpressions: []corev1.NodeSelectorRequirement{
+											{
```

Comment:
It seems this label is not applied to upgraded clusters. At least we need to document this.

Reply:
After looking into it, it seems that in some cases this label can be missing.
In fact, I've initially implemented the PR in a way that it requires either a CP or a Master label, but @fossedihelm [asked to](https://github.com/kubevirt/kubevirt/pull/11659#discussion_r1555270580) remove it.

@fossedihelm - IMHO we can be on the safe side and add both labels. TBH I don't think there's anything to lose from using such approach. Do you have any objections?

Reply:
Kinda weird! But agree to cover also this! Sorry for the wrong suggestion

Reply:
>- Introduce a new label "node-role.kubernetes.io/control-plane" that will be applied in parallel to "node-role.kubernetes.io/master" until the removal of the "node-role.kubernetes.io/master" label.
> - Make "kubeadm upgrade apply" add the "node-role.kubernetes.io/control-plane" label on existing nodes that only have the "node-role.kubernetes.io/master" label during upgrade.

From 1.20 k8s changelog https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.20.md :thinking:

Reply:
Asking to understand, why does Kubevirt should support the Master label if its deprecated on 1.24 while Kubevirt support k8s 1.29,8,7 at the moment?

Reply:
Why not? :)

As said above, in some environments the master label is still applicable (e.g. due to upgrade bugs). We have no real reason to avoid being on the safe side IMO.